catalog/core/catalog-core-migratable/src/main/java/org/codice/ddf/catalog/migratable/impl/MigrationTaskManager.java

```diff
@@ -0,0 +1,191 @@
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.catalog.migratable.impl;
+
+import java.io.File;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.codice.ddf.migration.MigrationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+
+import ddf.catalog.data.Result;
+
+/**
+ * Encapsulates all threading state for the catalog migration process and allows inspection of
+ * its progress through a public API and the Guava asynchronous call-backs.
+ *
+ * When using this task manager, always call exportSetup() first to ensure the target directory
+ * for metacards has been created. Failure to do so may cause a {@link MigrationException}.
+ *
+ * Repeat calls to exportMetacardQuery({@link List<Result>} results, {@link Long} exportGroupCount)
+ * as needed to add file write tasks to the {@link java.util.concurrent.BlockingQueue} of the
+ * {@link ExecutorService}. When finished, a call to exportFinish() is required to guarantee that
+ * no errors occurred and that the executor will be shutdown properly.
+ *
+ */
+class MigrationTaskManager {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MigrationTaskManager.class);
+
+    private final ListeningExecutorService taskExecutor;
+
+    private final AtomicReference<Throwable> taskFailure;
+
+    private final MigrationFileWriter fileWriter;
+
+    private final CatalogMigratableConfig catalogConfig;
+
+    /**
+     * The creation of a task manager requires the passing of migratable configurations so that
+     * task behavior can be explicitly defined (i.e. where to export, how many metacards per file,
+     * and other useful data).
+     *
+     * @param config The set of configuration options to use.
+     */
+    public MigrationTaskManager(final CatalogMigratableConfig config,
+            final MigrationFileWriter fileWriter, final ExecutorService executor) {
+        this.taskExecutor = MoreExecutors.listeningDecorator(executor);
+        this.taskFailure = createAtomicReference();
+        this.catalogConfig = config;
+        this.fileWriter = fileWriter;
+    }
+
+    /**
+     * Must be called prior to exportFinish and exportMetacardQuery to initialize the
+     * export directory.
+     *
+     * @throws MigrationException thrown if the setup fails
+     */
+    public void exportSetup() throws MigrationException {
```

Comment:
Instead of having an `exportSetup()` method that exists only to call `fileWriter.init()` to create the export directory, we could just create the directory in the `MigrationTaskManager`'s constructor and get rid of the `exportSetup()` and `fileWriter.init()` methods. This would prevent the clients from forgetting to call the `exportSetup()` method and put the object in an invalid state. The other alternative would be to add a check in the `exportMetacardQuery()` to make sure the directory actually exists before submitting the `Callable`s. Thoughts?